Documentation/ABI/TestContent.md

```diff
@@ -0,0 +1,178 @@
+# Runtime-discoverable test content
+
+<!--
+This source file is part of the Swift.org open source project
+
+Copyright (c) 2024 Apple Inc. and the Swift project authors
+Licensed under Apache License v2.0 with Runtime Library Exception
+
+See https://swift.org/LICENSE.txt for license information
+See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+-->
+
+This document describes the format and location of test content that the testing
+library emits at compile time and can discover at runtime.
+
+> [!WARNING]
+> The content of this document is subject to change pending efforts to define a
+> Swift-wide standard mechanism for runtime metadata emission and discovery.
+> Treat the information in this document as experimental.
+
+## Basic format
+
+Swift Testing stores test content records in a dedicated platform-specific
+section in built test products:
+
+| Platform | Binary Format | Section Name |
+|-|:-:|-|
+| macOS, iOS, watchOS, tvOS, visionOS | Mach-O | `__DATA_CONST,__swift5_tests` |
+| Linux, FreeBSD, Android | ELF | `swift5_tests` |
+| WASI | Statically Linked | `swift5_tests`[^1] |
+| Windows | PE/COFF | `.sw5test`[^2] |
+
+[^1]: SwiftWasm effectively uses the ELF format for its images, however it is
+      currently always statically linked and runtime discovery is performed
+      using a different mechanism than what we use on other ELF-based platforms.
+[^2]: On Windows, the Swift compiler [emits](https://github.com/swiftlang/swift/blob/main/stdlib/public/runtime/SwiftRT-COFF.cpp)
+      leading and trailing padding into this section, both zeroed and of size
+      `MemoryLayout<UInt>.stride`. Code that walks this section can safely skip
+      over this padding.
+
+### Record headers
+
+Regardless of platform, all test content records created and discoverable by the
+testing library have the following layout:
+
+```swift
+typealias TestContentRecord = (
+  kind: UInt32,
+  version: UInt16,
+  reserved1: UInt16,
+  accessor: (@convention(c) (_ outValue: UnsafeMutableRawPointer, _ hint: UnsafeRawPointer?) -> CBool)?,
+  context: UInt,
+  reserved2: UInt
+)
+```
+
+This type has natural size, stride, and alignment. Its fields are native-endian.
+If needed, this type can be represented in C as a structure:
+
+```c
+struct SWTTestContentRecord {
+  uint32_t kind;
+  uint16_t version;
```

Comment:
Is the version per-kind or per-record? I think that if `version` is per-kind, this is just a differently named TLV record. If it is per-record, then it constrains the `kind` field as well and should really precede the `kind` field.

Reply:
Thinking about it more, we probably don't need the `version` field. 32 bits is a huge namespace for record kinds and can support basic versioning by just changing the record kind. to something new.

The bytes are still going to sit unused (extending `kind` to 64 bits would definitely be a waste of bytes) but the reserved field can just be 32 bits wide until we find something to use them for.